Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift

```diff
@@ -80,7 +80,9 @@ public struct DeclarationRenderSection: Codable, Equatable {
         public let text: String
         /// The token programming kind.
         public let kind: Kind
-        
+        /// Any child tokens, used for `highlightDiff`-type tokens.
+        public let tokens: [Token]?
```

Comment:
I don't like this way of modeling groups of highlighted tokens. I find it very unexpected that `Token` is recursive and I find that it doesn't fit the documented meaning of a this type:
> A lexical token is a string with an associated meaning in source code.

If this was only a public API we'd have a way to deprecate it and refine in the future. Unfortunately, we don't have a similar process for making breaking changes to the Render Node spec—which also models "DeclarationToken" with recursion. This means that if we make this change now, it's practically final. 

I'm not ready to commit to this design yet. It would be discuss it offline with the team to see how other people feel about it.

Here's two alternatives that I feel are worth considering:

- Model highlighting as a property of an individual token, for example
  ```diff
   public struct Token: Codable, Hashable, Equatable {
       public let text: String
       public let kind: Kind
  +    public let highlighted: Bool
   }
  ```
  
- Model highlighting as ranges of indices, parallel to a list of tokens, for example
    ```diff
  +public struct HighlightedTokens {
  +    let tokens: [Token]
  +    let highlights: [Range<Int>]? = nil
  +}
  ```

Reply:
cc @mportiz08 

Swift-DocC-Render already handles this model of tokens, with a partial support for rendering `added` and `removed` token spans that look like this. My original version of this PR added a boolean `highlightDiff` flag to the tokens (similar to your option 1), but it complicated the Swift-DocC-Render integration due to needing to add a style to all the distinct token kinds to make the rendering work. If Marcus is fine with using that version of the Render JSON for the final spec, i can revert the commit that added the recursive property.

Reply:
I think it makes more sense to model recursively as a tree of tokens since that is how it gets rendered and fits in very nicely with the existing way that change highlights are already rendered. My main concern is that adding a new additional boolean flag to every single token in the JSON seems like a very inefficient design for this very small/focused use case as opposed to just introducing a new token type that only applies for this situation with presenting multiple overloads in one render node.

However, if this doesn't work for the way that tokens are represented in DocC itself, of course we can always try to work with some other JSON format.

I'm not very familiar with this code, but it sounds like more of an issue with how this data is modeled in the Swift code (using a single static type for all tokens and optional properties for certain kinds like this and linkable tokens as opposed to modeling this in a generic way where different tokens can have different data).

Let me know what everyone decides on, and I can try to update the renderer PR to handle the new design. :)

Reply:
I think I have to agree with @d-ronnqvist , although changing this might introduce more work in DocC Render. It is a bit difficult to follow the code in `translateDeclaration`, for example in the various conditions similar to this one:

```
} else if previousToken.kind != .highlightDiff,
    token.kind == .highlightDiff,
    token.tokens?.count == 1,
    let innerToken = token.tokens?.first,
    innerToken.kind == .text,
    innerToken.text.first?.isWhitespace == true
```

`innerToken` seems like an odd name. And `token.tokens?.first` is surprising.

I'm also concerned about the possibility of other renderers that consume DocC Render JSON  that might not anticipate the concept of recursive tokens. I'd go with @d-ronnqvist 's first suggestion, which sounds like what @QuietMisdreavus had originally. Maybe the RenderJSON could add a new, optional boolean flag - something like this:

```
{
  "text": ": ",
  "kind": "text"
},
{ 
  "preciseIdentifier": "some-id",
  "text": "some-text",
  "highlightDiff": true,
  "kind": "typeIdentifier",
  "identifier": "other-id"
},
{ 
  "kind": "text",
  "text": "?) -> "
},
```

So in this example, the two tokens with no `highlighted` flag at all assume this is false and would render normally. The one highlighted token only would have `"highlighted": true`

Reply:
> My main concern is that adding a new additional boolean flag to every single token in the JSON seems like a very inefficient design for this very small/focused use case

I don't buy this argument. You are opposed to adding a boolean to every token but have no concern with adding an array to every token. In both cases we would only encode a value if it's not the default.

> as opposed to just introducing a new token type that only applies for this situation with presenting multiple overloads in one render node.

This is _very_ far from "just introducing a new token type". It is fundamentally redefining the structure of a token from a single element in a list to an arbitrary hierarchy.

Reply:
> > My main concern is that adding a new additional boolean flag to every single token in the JSON seems like a very inefficient design for this very small/focused use case
> 
> I don't buy this argument. You are opposed to adding a boolean to every token but have no concern with adding an array to every token. In both cases we would only encode a value if it's not the default.

This is only because of how the Swift code is structured. In the JSON, you will only ever encounter the array given the right kind value. I mentioned that I can see how this is problematic with how the Swift code is structured and could be solved by using a more generic type for these tokens, but I'm sure it's difficult to make that change at this point.

> > as opposed to just introducing a new token type that only applies for this situation with presenting multiple overloads in one render node.
> 
> This is _very_ far from "just introducing a new token type". It is fundamentally redefining the structure of a token from a single element in a list to an arbitrary hierarchy.

Agreed. I assumed that it was already tree based, since we support this kind of token for other features, but those are probably features that DocC doesn't support itself yet. I think the tree format gives us much more flexibility for doing multiple kinds of highlighting for a given run of text, but it sounds like DocC is pretty tied to the flat list for tokens, so I can understand if you'd prefer to stick with that.

Reply:
Part of me wonders if we could square this circle by splitting the token type in Swift code into an enum, separating "tokens" from "token spans":

```swift
struct Token {
  // ... the existing token data ...
}

enum RenderToken {
  case token(Token)
  case tokenSpan(kind: TokenSpanKind, tokens: [Token])
}

enum TokenSpanKind {
  case highlightDiff
  // future cases `added` and `removed` if/when we support API diffing
}
```

This can then render into JSON the way that Swift-DocC-Render expects: `case token` gets all its fields emitted in-place, and `case tokenSpan` renders `kind` and `tokens` as if the `TokenSpanKind`s were part of the original token kinds. This way we separate the concerns in Swift but continue to integrate with DocC-Render the way it expects.

Reply:
This doesn't solve the issue of legibility in `translateDeclaration`, but at least it helps create a separation of concerns that validates that only some "token kinds" will have token lists.

Reply:
Using two distinct types is better but it doesn't address the issue that changing the token structure from a list to a hierarchy will break existing code.

For example, in DocC alone (ignoring all the usage in unit tests) we have:
- both the navigator index and link diagnostics use `tokens.map(\.text).joined()` to computes a plain text version of a declaration for display.
- `subHeading(for:symbolTitle:symbolKind:)` uses `tokens.firstIndex(of: initKeyword)` and indexed subscripting to rewrite the declaration.
- `mapNameFragmentsToIdentifierKind(matching:)` uses `count`, `removeFirst()`, and indexed subscripting to rewrite the declaration.

This existing code will silently produce incorrect results if declaration tokens are changed from a list of simple elements to a hierarchy. We could fix such bugs in DocC itself but this is public API so clients would need to do the same.

Reply:
> I'm not very familiar with this code, but it sounds like more of an issue with how this data is modeled in the Swift code (using a single static type for all tokens and optional properties for certain kinds like this and linkable tokens as opposed to modeling this in a generic way where different tokens can have different data).

The issue is primarily with how this is modeled in the RenderNode spec—a single object type with a mix of required and optional properties—and secondary with how the Swift code mirrors the RenderNode spec.

> > > My main concern is that adding a new additional boolean flag to every single token in the JSON seems like a very inefficient design for this very small/focused use case
> >
> >
> > I don't buy this argument. You are opposed to adding a boolean to every token but have no concern with adding an array to every token. In both cases we would only encode a value if it's not the default.
>
> This is only because of how the Swift code is structured. In the JSON, you will only ever encounter the array given the right kind value. I mentioned that I can see how this is problematic with how the Swift code is structured and could be solved by using a more generic type for these tokens, but I'm sure it's difficult to make that change at this point.

Again, the Swift mirrors the RenderNode spec which defines a single "DeclarationToken" object with 2 required and 2 non-required properties.

> > > as opposed to just introducing a new token type that only applies for this situation with presenting multiple overloads in one render node.
> >
> >
> > This is _very_ far from "just introducing a new token type". It is fundamentally redefining the structure of a token from a single element in a list to an arbitrary hierarchy.
>
> Agreed. I assumed that it was already tree based, since we support this kind of token for other features, but those are probably features that DocC doesn't support itself yet. I think the tree format gives us much more flexibility for doing multiple kinds of highlighting for a given run of text, but it sounds like DocC is pretty tied to the flat list for tokens, so I can understand if you'd prefer to stick with that.

I'm not sure what you are referring to when you say "we support this kind of token for other features" here but it doesn't seem to be about declaration tokens. I can't find anywhere where declaration tokens ("DeclarationToken" in the RenderNode spec) is tree-based. Similarly, "EndpointToken" isn't tree-based and every single "tokens" property in the RenderNode spec is defined as an array of either of these two types.

In my opinion the RenderNode spec correctly models tokens as a simple list. A "token" in the context of syntax/code/declarations is a term-of-art and it has specific meaning which isn't recursive. If it was, then compilers/tokenizers would model their tokens as a hierarchy.

Reply:
> I'm not sure what you are referring to when you say "we support this kind of token for other features" here but it doesn't seem to be about declaration tokens.

I believe this is about Swift-DocC-Render, which handles different [token kinds](https://github.com/apple/swift-docc-render/blob/20e5ea36fb908d1b25dcd009cb4e5cf01957c907/src/components/DocumentationTopic/PrimaryContent/DeclarationToken.vue#L18-L32) and [fields](https://github.com/apple/swift-docc-render/blob/20e5ea36fb908d1b25dcd009cb4e5cf01957c907/src/components/DocumentationTopic/PrimaryContent/DeclarationToken.vue#L82-L105) than Swift-DocC currently emits. It already handles the recursive `tokens` property that i added here, because it has support for rendering `added` and `removed` token spans (though some of the requisite CSS classes/variables are missing, likely because of the lack of Swift-DocC support).

I imagine if we want to add support for rendering differences in declarations (e.g. leveraging the existing `RenderJSONDiffable` protocol and machinery), then we'll need to reckon with these recursive tokens anyway, or change Swift-DocC-Render to handle them differently than what it's currently expecting.

Reply:
> I imagine if we want to add support for rendering differences in declarations (e.g. leveraging the existing `RenderJSONDiffable` protocol and machinery), then we'll need to reckon with these recursive tokens anyway, or change Swift-DocC-Render to handle them differently than what it's currently expecting.

If we want to support different types of highlights then that seems like it would be nicely solved by an enumeration on the token.

```diff
 public struct Token: Codable, Hashable, Equatable {
     public let text: String
     public let kind: Kind
     public let identifier: String?
+    public let highlight: Highlight?
+    public enum Highlight: String, Codable {
+        case changed
+        case added
+        case removed
+    }
 }
```

If we only need one highlight type for now—for the "highlight diff" behavior—we could start with only a `changed` case and add `added` and `removed` cases when we need to represent those types of highlights.

Reply:
I'm sorry, I think I must have misunderstood. Is this a 3rd option different from your original design with the highlight boolean flag? We can make pretty much anything work—I just need some clarification on the intended schema. Would this be similar but with a constant "changed" string or something for the flag? Just making sure I understand.

Reply:
Right, in this case it would be a constant `changed` string, with the intent of folding the `added` and `removed` token types that Swift-DocC-Render currently handles as options here in the future, rather than representing them as recursive tokens.

Reply:
Updated: https://github.com/apple/swift-docc-render/pull/847/commits/e549cefd30e3a2e6e7ba57f9d7ec1e397baa4fa4

Let me know when the DocC side is updated, and I can do some quick tests.